java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java

```diff
@@ -95,6 +103,37 @@ public void visitMethod(MethodTree tree) {
     }
   }
 
+  /**
+   * Method's return type requires type parameter and therefore cannot be made static.
+   */
+  private static boolean returnRequiresParentTypeParameter(Symbol.MethodSymbol symbol) {
+    // called from visitMethod, so we have declaration (NPE not possible)
+    ClassTree parent = (ClassTree) symbol.declaration().parent();
+    Set<Symbol> parentTypeParam =
+      parent
+        .typeParameters()
+        .stream()
+        .map(TypeParameterTree::symbol)
+        .collect(Collectors.toUnmodifiableSet());
+    Symbol.TypeSymbol returnType = symbol.returnType();
+    List<Type> returnTypeVars = new ArrayList<>();
+    collectTypeVars(returnTypeVars, returnType.type());
+    return returnTypeVars
+      .stream()
+      .map(Type::symbol)
+      .anyMatch(parentTypeParam::contains);
+  }
+
+  private static void collectTypeVars(List<Type> accumulator, Type t) {
+    for(Type tt: t.typeArguments()) {
+      if(tt.isTypeVar()) {
+        accumulator.add(tt);
+      } else {
+        collectTypeVars(accumulator, tt);
```

Comment:
Is there a way to document, using a comment or casting it into a new variable, what type `tt` could be apart from a `TypeVar`? No need to be exhaustive but an example would be nice